crates/red_knot_test/README.md

```diff
@@ -0,0 +1,430 @@
+
+```py
+from typing import reveal_type
+
+reveal_type("foo")  # revealed: Literal["foo"]
+```
+
+For convenience, type checkers also pretend that `reveal_type` is a built-in, so that this import is
+not required. Using `reveal_type` without importing it issues a diagnostic warning that it was used
+without importing it, in addition to the diagnostic revealing the type of the expression.
+
+The `# revealed:` assertion must always match a revealed-type diagnostic, and will also match the
+undefined-reveal diagnostic, if present, so it's safe to use `reveal_type` in tests either with or
+without importing it. (Style preference is to usually not import it in tests, unless specifically
+testing something about the behavior of importing it.)
+
+#### error
+
+A comment beginning with `# error:` is an assertion that a type checker diagnostic will
+be emitted, with text span starting on that line. If the comment is simply `# error:`, this will
+match any diagnostic. The matching can be narrowed in three ways:
+
+- `# error: 8` requires that the matched diagnostic's text span begins on column 8 (one-indexed) of
+    this line.
+- `# error: [invalid-assignment]` requires that the matched diagnostic have the rule code
+    `invalid-assignment`.
+- `# error: "Some text"` requires that the matched diagnostic's full message contain the text `Some text`.
+
+Any combination of two or all three of these can be used, but they must go in order: first column,
+if present; then rule code, if present, then contains-text, if present. For example, an assertion
+using all three would look like `# error: 8 [invalid-assignment] "Some text"`.
+
+Error assertions in tests intended to test type checker semantics should use primarily rule-code
+assertions, with occasional contains-text assertions where needed to disambiguate.
+
+Some tests may exist in order to assert on the correct text-range or text of a diagnostic; these
+should use the appropriate assertions.
+
+### Assertion locations
+
+An assertion comment may be a line-trailing comment, in which case it applies to the line it is on:
+
+```py
+x: str = 1  # error: [invalid-assignment]
+```
+
+Or it may be a comment on its own line, in which case it applies to the next non-assertion-comment
+line:
+
+```py
+# error: [invalid-assignment]
+x: str = 1
+```
+
+Multiple assertions applying to the same line may be stacked:
+
+```py
+# error: [invalid-assignment]
+# revealed: Literal[1]
+x: str = 1
+```
+
+Intervening empty lines or non-assertion comments are not allowed; an assertion stack must be one
+assertion per line, immediately following each other, with the line immediately following the last
+assertion the line of source code emitting the matched diagnostics.
+
+## Multi-file tests
+
+Some tests require multiple files, with imports from one file to another. Multiple fenced code
+blocks represent multiple embedded files. Since files must have unique names, only zero or one files
```

Comment:
Should it be "only one file can use..." instead of "only zero or one files can use..."?